olp-cpp-sdk-core/include/olp/core/thread/Continuation.h

```diff
@@ -0,0 +1,277 @@
+  using FailedCallback = std::function<void(client::ApiError)>;
+  /// Type of return value of Continuation Task.
+  using OutResultType = std::unique_ptr<UntypedSmartPointer>;
+
+  /// Type of Continuation Type.
+  using TaskType = std::function<OutResultType(void*)>;
+  /// Default callback type, uses void* as a parameter which is later converted
+  /// to the specific type, callback doesn't have a return type.
+  using CallbackType = std::function<void(void*)>;
+  /// An internal type of tasks in Continuation.
+  using AsyncTaskType = std::function<void(void*, CallbackType)>;
+  /// An alias for a processing tasks finalization type.
+  using FinalCallbackType = std::function<void(void*, bool)>;
+  /// An alias of a pair of Task Continuation types.
+  using ContinuationTask = std::pair<AsyncTaskType, TaskType>;
+
+  /// A default constructor of `ContinuationImpl`.
+  ContinuationImpl() = default;
+
+  /**
+   * @brief A contructor for a continuation using the given parameters.
+   *
+   * @param task_scheduler The `TaskScheduler` instance.
+   * @param context The `ExecutionContext` instance.
+   * @param task The `ContinuationTask` instance.
+   */
+  ContinuationImpl(std::shared_ptr<TaskScheduler> task_scheduler,
+                   ExecutionContext context, ContinuationTask task);
+
+  /**
+   * @brief Adds an asynchronous task (one taking a callback) as the next
+   * task of the `ContinuationImpl`.
+   *
+   * @param task A task for adding to the continuation chain.
+   *
+   * @return The `ContinuationImpl` instance.
+   */
+  ContinuationImpl Then(ContinuationTask task);
+
+  /**
+   * @brief Starts the execution of a task continuation chain.
+   *
+   * @param callback Handles the finalization of a task chain.
+   */
+  void Run(FinalCallbackType callback);
+
+  /**
+   * @brief Provides the execution context object.
+   *
+   * @return The `ExecutionContext` instance.
+   */
+  const ExecutionContext& GetExecutionContext() const;
+
+  /**
+   * @brief Provides cancellation context object.
+   *
+   * @return The `CancellationContext` instance.
+   */
+  const client::CancellationContext& GetContext() const;
+
+  /**
+   * @brief Provides a status whether CancellationContext is cancelled.
+   *
+   * @return True if Continuation is cancelled; false otherwise.
+   */
+  bool Cancelled() const;
+
+  /**
+   * @brief Sets a callback on calling SetError.
+   *
+   * @param callback Handles an execution finalization of a continuation chain.
+   */
+  void SetFailedCallback(FailedCallback callback);
+
+  /**
+   * @brief Clears the continuation chain tasks.
+   */
+  void Clear();
+
+ private:
+  std::shared_ptr<thread::TaskScheduler> task_scheduler_;
+  std::deque<ContinuationTask> tasks_;
+  ExecutionContext execution_context_;
+
+  /// a variable shows whether it is allow to do the changes to the task chain,
+  /// used to prevent using methods of `Continuation` after it was run.
+  bool continuation_change_allowed{true};
```

Comment:
Or during it runs. Can the user simply call Run() on the continuation and then call Then() without you crashing?

```suggestion
  bool change_allowed_{true};
```

Reply:
renamed the variable.
Added a unit test which checks that Then, Finally and Run don't cause an exceptions
and aren't called if methods were already called once.